Re-enable group tests 1/?

raphaelrobert opened the pull request:
This PR re-enables the previously commented out group tests

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -6,6 +6,7 @@ use maelstrom::key_packages::*;
 #[test]
 fn create_commit_optional_path() {
+    use maelstrom::extensions::*;
     let ciphersuite_name = CiphersuiteName::MLS10_128_DHKEMX25519_AES128GCM_SHA256_Ed25519;
```
Tests should cover all supported ciphersuites.

franziskuskiefer replied:
blocked by #168

franziskuskiefer replied:
#168 got merged so we should add tests for all ciphersuites.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -17,97 +18,149 @@ fn create_commit_optional_path() {
     );
+    let epoch_proposals = vec![bob_add_proposal];
+    let (mls_plaintext_commit, _welcome_bundle_alice_bob_option) = match group_alice_1234
+        .create_commit(group_aad, alice_signature_key, epoch_proposals, true)
```
This more about the API. But having boolean arguments, especially in public APIs, is not great. It's never clear what the `true` here means. So for this API this should be something like `true /* force self-update */`. But maybe we should consider adding a separate function with this set to `true`, like `create_commit_self_update`.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -17,97 +18,149 @@ fn create_commit_optional_path() {
-        false,
-    ) {
+    let epoch_proposals = vec![bob_add_proposal];
+    let (mls_plaintext_commit, welcome_bundle_alice_bob_option) = match group_alice_1234
```
So we add Bob twice to the same group? Shouldn't this fail?

raphaelrobert replied:
We only create a proposal/commit that we don't apply. This should not fail.

franziskuskiefer replied:
Then there are comments missing explaining what this does (as mentioned in another comment already).

franziskuskiefer replied:
🤔

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -159,109 +212,156 @@ fn basic_group_setup() {
 
-    assert_eq!(group_alice.tree.nodes, group_bob.tree.nodes);
-    group_alice.tree.print(&format!("\n{:?}", ms1));
+    assert_eq!(
```
This looks like a pretty crucial test. It should probably get an explanation.
Also, the `_print_tree` should only be interesting if this fails. But it won't be reached in the case the assert fails.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -159,109 +212,156 @@ fn basic_group_setup() {
 
-/*
 #[test]
 fn group_operations() {
```
What operations?
This needs a description of what's exactly happening/tested in here.

franziskuskiefer replied:
🤔

raphaelrobert replied:
That was added, but I think you are looking at the wrong file

franziskuskiefer reviewed (changes requested):
So this PR should enable the group tests. It's to expect that changes to the code are needed for this. But there's no explanation to which and why. So I only really looked at the test for now.
The description in the PR also doesn't appear to be correct as it only enables a part of the commented out test.

raphaelrobert commented:
> So this PR should enable the group tests. It's to expect that changes to the code are needed for this. But there's no explanation to which and why. So I only really looked at the test for now.
> The description in the PR also doesn't appear to be correct as it only enables a part of the commented out test.

This PR indeed only covers a part of the tests. Enabling more of them would make the PR grow beyond the limit we imposed ourselves.
This PR fixes all the obvious problems that were introduced in the last weeks and that cause the inner workings to break.
There will be at least one follow-up PR that will close #77.

franziskuskiefer commented on src/config.rs:
```diff
@@ -52,6 +52,18 @@ pub struct Config {
 }
 
+impl Config {
+    pub fn get_protocol_version_ref(&self) -> &[ProtocolVersion] {
```
Do we need these? The global config can be accessed already with the functions below.
Also, if we want these functions we should put them in the same `impl` block.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -17,97 +18,149 @@ fn create_commit_optional_path() {
+    let bob_signature_key = &bob_identity.get_signature_key_pair().get_private_key();
+
+    // Mandatory extensions
+    let capabilities_extension = Box::new(CapabilitiesExtension::default());
```
🤔

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -17,97 +18,149 @@ fn create_commit_optional_path() {
+
+    // Mandatory extensions
+    let capabilities_extension = Box::new(CapabilitiesExtension::default());
+    let lifetime_extension = Box::new(LifetimeExtension::new(60));
```
🤔

raphaelrobert replied:
This is something for #164 then, we don't want to do that here.

franziskuskiefer reviewed (changes requested):
Maybe I'm missing something but I don't find a description of the changes.

franziskuskiefer reviewed (changes requested):
I still can't find a description of the changes.

raphaelrobert commented:
> I still can't find a description of the changes.

The last commit only addresses your comments, as the name suggests. There are no other changes.

franziskuskiefer commented:
> > I still can't find a description of the changes.
> 
> The last commit only addresses your comments, as the name suggests. There are no other changes.

There is no description of any changes in this PR. That's what I'm waiting for before reviewing. The description is "This PR re-enables the previously commented out group tests". But there are a ton of code changes in this PR. They all need a high level description here.

raphaelrobert commented:
> > > I still can't find a description of the changes.
> > 
> > 
> > The last commit only addresses your comments, as the name suggests. There are no other changes.
> 
> There is no description of any changes in this PR. That's what I'm waiting for before reviewing. The description is "This PR re-enables the previously commented out group tests". But there are a ton of code changes in this PR. They all need a high level description here.

The context of your question wasn't clear.
There is a high-level description which you didn't quote, but here is a more detailed one:
 - Re-enable part of the group tests to test basic operations like Add & Update
 - Validate public keys from the UpdatePath
 - Change return type of KeyPackage verification from boolean to more specific errors
 - Fix direct path calculation in treemath
 - Fix missing RatchetTree clone in create_commit
 - Refactor part of PrivateTree
 - Fix various smaller issues that made the tests break

franziskuskiefer commented on src/tree/private_tree.rs:
```diff
@@ -134,13 +134,13 @@ impl PrivateTree {
         let hash_len = ciphersuite.hash_length();
         let start_secret = match start_secret {
-            Some(secret) => hkdf_expand_label(ciphersuite, secret, "path", &[], hash_len),
+            Some(secret) => secret.to_vec(),
```
This appears to be wrong. According to spec we have `path_secret[0] = DeriveSecret(leaf_secret, "path")`. So unless you changed the way this function is called the `start_secret` shouldn't be used directly.

raphaelrobert replied:
You are right and that's described here: https://github.com/raphaelrobert/maelstrom/issues/165
It's already fixed in the follow-up PR.

franziskuskiefer commented on src/tree/treemath.rs:
```diff
@@ -136,7 +136,7 @@ pub(crate) fn dirpath_long(
 ) -> Result<Vec<NodeIndex>, TreeMathError> {
     let r = root(size);
     if index == r {
-        return Ok(vec![]);
```
Please add a test for this.

raphaelrobert replied:
This can only be tested with interop test vectors, recalculating the same thing is a moot point. We should do that during the IETF hackathon.

franziskuskiefer replied:
A simple test with predefined tree is certainly possible. That's all that's needed for this for now.
Of course we need more and better tests. But this code change should have a test. Same below.

raphaelrobert replied:
There's a test now in `treemath_test.rs`

franziskuskiefer commented on src/tree/treemath.rs:
```diff
@@ -156,7 +156,7 @@ pub(crate) fn direct_path_root(
 ) -> Result<Vec<NodeIndex>, TreeMathError> {
     let r = root(size);
     if index == r {
-        return Ok(vec![]);
```
And this one as well.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
+        MLSPlaintextContentType::Commit((commit, _)) => commit,
+        _ => panic!(),
+    };
+    assert!(commit.path.is_some());
```
Can we test more here than just the fact that there is a path?

raphaelrobert replied:
The follow-up PR does test more. We need the API change for that.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
-    // Alice updates
-    let alice_update_proposal = group_alice_1234.create_update_proposal(
+    // Alice adds Bob without forced self-update
+    // Since there are on Add Proposals, this not generate a path field on the Commit
```
"... only Add Proposals this does not generate ..."

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
-    ) {
-        Ok(c) => c,
-        Err(e) => panic!("Error creating commit: {:?}", e),
+    // Alice applies Commit
```
Which one? Looks like the second without forced self-update.

franziskuskiefer replied:
🤔

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
-        _ => panic!(),
+    // Bob creates group from Welcome
+    let mut group_bob_1234 = match MlsGroup::new_from_welcome(
+        welcome_bundle_alice_bob_option.unwrap(),
```
Shouldn't this `unwrap` be a test case earlier?

raphaelrobert replied:
Do you mean we should test if the welcome message exists? I can add that test case.

franziskuskiefer replied:
Yeah, we shouldn't just panic here when it's missing but check for this earlier.

raphaelrobert replied:
The check is now in L276

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
-        vec![alice_update_proposal],
-        true,
+        proposals.clone(),
+        false,
```
`/* force self update */`

raphaelrobert replied:
When you use rust-analyzer, you get that for free, that's why I don't make these kind of annotations.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -15,94 +16,146 @@ fn create_commit_optional_path() {
+        }
+    */
+
+    match group_bob_1234.apply_commit(commit_mls_plaintext, proposals, vec![]) {
```
Maybe use `.expect()`? There are a lot more like this in this file.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -150,109 +203,162 @@ fn basic_group_setup() {
 
-/*
 #[test]
+/// This test simulates various group operations like Add, Update, Remove in a small group
```
To make this more digestible having a high level description of what's supposed to happen would be nice. Something like
* Alice creates group
* Alice adds Bob
* Bob removes Alice 
...

raphaelrobert replied:
Good point. I'll add this at the end in the last PR, because the order is still very much in flux. These tests need a better system anyway, to be able to cover more cases. For now the idea is to just re-enable the tests that were already written.

franziskuskiefer replied:
Yeah, That's what @kkohbrok is currently working on (no PR yet).

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -150,109 +203,162 @@ fn basic_group_setup() {
+                group_aad,
+                &alice_credential_bundle,
+                epoch_proposals.clone(),
+                false, // TODO force update seems broken
```
Shouldn't the `create_commit_optional_path` test fail then?

raphaelrobert replied:
That comment was outdated

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -150,109 +203,162 @@ fn basic_group_setup() {
+        };
+        let commit = match &mls_plaintext_commit.content {
+            MLSPlaintextContentType::Commit((commit, _)) => commit,
+            _ => panic!(),
```
Adding a little more information to panics would help debugging later.

franziskuskiefer reviewed (changes requested):
Thanks for the update and additional description. See inline comments

Please also fix the clippy warning.

Other changes in here (that should co into the commit message in the end):
*  new public function `as_application_message` to get a `MLSPlaintext` as application message (byte slice).

raphaelrobert commented on src/group/mls_group/apply_commit.rs:
```diff
@@ -76,11 +76,14 @@ impl MlsGroup {
             }
             let serialized_context = self.group_context.encode_detached().unwrap();
-            if !mls_plaintext.verify(Some(serialized_context), kp.get_credential()) {
+            if !mls_plaintext.verify(Some(serialized_context.clone()), kp.get_credential()) {
```
https://github.com/raphaelrobert/maelstrom/issues/171

raphaelrobert commented on src/group/mls_group/mod.rs:
```diff
@@ -221,7 +221,8 @@ impl Api for MlsGroup {
         let mut roster = Vec::new();
         for i in 0..tree.leaf_count().as_usize() {
-            let node = &tree.nodes[i];
+            let leaf_index = LeafIndex::from(i);
```
The problem was that it was off by a factor of 2. The new indexing wasn't available when I wrote this, I'm adapting it now.

raphaelrobert commented on src/key_packages/mod.rs:
```diff
@@ -82,7 +86,8 @@ impl KeyPackage {
     /// * make sure that the lifetime is valid
-    pub(crate) fn verify(&self) -> bool {
+    /// Returns `Ok(())` if all checks succeed and `KeyPackageError` otherwise
+    pub fn verify(&self) -> Result<(), KeyPackageError> {
```
Agreed. The problem right now is that the public error code all have integer values which prevents the nesting. We should probably revert on that decision for now and do this only for FFI bindings.

raphaelrobert commented on src/tree/mod.rs:
```diff
@@ -49,7 +49,7 @@ mod test_treemath;
 mod test_util;
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
```
Yes, we have to make a complete copy of the tree, including the of the secrets. There might be a way around it, but it would require a crazy setup of references, and at this point I want to keep it simple.

franziskuskiefer replied:
Any change like this will cost us a couple weeks to fix down the road. If you think that's worth it, you can keep it as is.

raphaelrobert replied:
We can discuss this offline, but I don't see a big issue with copying the tree right now. This is also what the spec is suggesting.

raphaelrobert commented on src/tree/mod.rs:
```diff
@@ -338,11 +329,9 @@ impl RatchetTree {
 
-        debug_assert!(sender_direct_path.len() > common_path.len());
-        if sender_direct_path.len() <= common_path.len() {
+        if sender_direct_path.len() < common_path.len() {
```
No. the `debug_assert` didn't seem to make much sense, not sure how it ended up there.

franziskuskiefer replied:
Then we need a test 🙂

raphaelrobert replied:
I thought about this some more and concluded that this cannot occur with malformed payload. It must be a library error on the treemath level and I changed the check accordingly.

raphaelrobert commented:
Thanks for all the feedback @franziskuskiefer!

I addressed it in the last two commits, or through comments.

Unfortunately P256 is still broken so it didn't make it in here. I need to investigate what it is this time.

franziskuskiefer commented on tests/test_group.rs:
```diff
@@ -272,11 +272,12 @@ fn group_operations() {
             _ => panic!("Wrong content type"),
         };
         assert!(commit.path.is_none());
+        // Check that the fucntion returned a Welcome message
```
"function"

franziskuskiefer reviewed (changes requested):
Almost there.

There are always a lot of unresolved and unanswered comments. This costs a lot of time to go through every time and increases the number of roundtrips we need to get this reviewed.

franziskuskiefer reviewed (approved):
Ok lets get this in and address everything that's still open later.
Please make sure to squash before merging and make the commit message cover all changes in here.
